src/centengine/engine_impl.cc

```diff
@@ -792,6 +793,130 @@ grpc::Status engine_impl::RemoveServiceAcknowledgement(
   return grpc::Status::OK;
 }
 
+grpc::Status engine_impl::AcknowledgementHostProblem(
+    grpc::ServerContext* context __attribute__((unused)),
+    const EngineAcknowledgement* request,
+    CommandSuccess* response) {
+  auto fn = std::packaged_task<int32_t(void)>([request]() -> int32_t {
+    std::shared_ptr<engine::host> temp_host;
+    /* get the host */
+    auto it = host::hosts.find(request->host_name());
+    if (it != host::hosts.end())
+      temp_host = it->second;
+    if (temp_host == nullptr)
+      return 1;
+    /* cannot acknowledge a non-existent problem */
+    if (temp_host->get_current_state() == host::state_up)
+      return 1;
+    /* set the acknowledgement flag */
+    temp_host->set_problem_has_been_acknowledged(true);
+    /* set the acknowledgement type */
+    if (EngineAcknowledgement::Type_Name(request->type()) == "STICKY")
+      temp_host->set_acknowledgement_type(ACKNOWLEDGEMENT_STICKY);
+    else
+      temp_host->set_acknowledgement_type(ACKNOWLEDGEMENT_NORMAL);
+    /* schedule acknowledgement expiration */
+    time_t current_time(time(nullptr));
+    temp_host->set_last_acknowledgement(current_time);
+    temp_host->schedule_acknowledgement_expiration();
+    /* send data to event broker */
+    char* ack_author = strdup(request->ack_author().c_str());
+    char* ack_data = strdup(request->ack_data().c_str());
+    broker_acknowledgement_data(NEBTYPE_ACKNOWLEDGEMENT_ADD, NEBFLAG_NONE,
+                                NEBATTR_NONE, HOST_ACKNOWLEDGEMENT,
+                                static_cast<void*>(temp_host.get()), ack_author,
+                                ack_data, request->type(), request->notify(),
+                                request->persistent(), nullptr);
+    /* send out an acknowledgement notification */
+    if (request->notify())
+      temp_host->notify(notifier::reason_acknowledgement, ack_author, ack_data,
+                        notifier::notification_option_none);
```

Comment:
here, notify() needs a std::string on ack_author and ack_data.
If you give a char*, it will be changed into a std::string, so here are two new copies.
The solution, use request->ack_author() and request->ack_data() directly.